pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to apply adjoint jacobian differentiation"""
+
+import warnings
+import numpy as np
+
+import pennylane as qml
+
+from pennylane.operation import operation_derivative
+from pennylane.tape import QuantumTape
+from pennylane.measurements import ExpectationMP
+
+from .apply_operation import apply_operation
+
+from .initialize_state import create_initial_state
+
+# pylint: disable=protected-access, too-many-branches
+
+
+def _dot_product_real(bra, ket, num_wires):
+    """Helper for calculating the inner product for adjoint differentiation."""
+    # broadcasted inner product not summing over first dimension of the bra tensor
+    sum_axes = tuple(range(1, num_wires + 1))
+    return qml.math.real(qml.math.sum(qml.math.conj(bra) * ket, axis=sum_axes))
+
+
+def adjoint_jacobian(
+    tape: QuantumTape, prep_operation: qml.operation.StatePrep = None
```

Comment:
The prep operation should be accessible from `tape._prep`, and therefore doesn't need to be a separate argument.